### PR TITLE
Remove dependencies from TaskExecutionSpecification

### PR DESCRIPTION
**Pull request by edoakes**
## What do these changes do?

Removes the `dependencies` field from `TaskExecutionSpecification` in favor of a new `PreviousActorTaskDummyObjectId` field in `TaskSpecification`.

This was an artifact left over from a previous design, and the dependencies were never actually updated during runtime. Rather, it was only used to specify dependencies for actor tasks on the previously run actor task. Now, there is a dedicated field for this use case that is only specified as part of actor task submissions.

The change has a nice byproduct of simplifying the task submission interfaces in the core worker.

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by edoakes** [comment]
Note that this touches many of the same files as https://github.com/ray-project/ray/pull/5147

**Comment by edoakes** [comment]
Updated to rebase off of the changes merged in #5147

File: src/ray/core_worker/context.cc [hunk 1]
```diff
@@ -20,7 +20,9 @@ struct WorkerThreadContext {
     put_index = 0;
   }
 
-  void SetCurrentTask(const TaskSpecification &spec) { SetCurrentTask(spec.TaskId()); }
+  void SetCurrentTask(const TaskSpecification &task) {
```
Comment (stephanie-wang) [inline src/ray/core_worker/context.cc]: I would rename these arguments `task_spec` to make it clear that it is not a `Task` (which includes runtime information). Same elsewhere.

File: src/ray/core_worker/task_interface.cc [hunk 2]
```diff
@@ -177,11 +175,10 @@ Status CoreWorkerTaskInterface::SubmitActorTask(ActorHandle &actor_handle,
       ObjectID::FromBinary(actor_handle.ActorID().Binary());
   builder.SetActorTaskSpec(actor_handle.ActorID(), actor_handle.ActorHandleID(),
                            actor_creation_dummy_object_id,
+			   actor_handle.ActorCursor(), // previous_actor_task_dummy_object_id
```
Comment (stephanie-wang) [inline src/ray/core_worker/task_interface.cc]:
```suggestion
			   /*previous_actor_task_dummy_object_id=*/actor_handle.ActorCursor(),
```

File: python/ray/includes/task.pxi [hunk 3]
```diff
@@ -229,6 +231,13 @@ cdef class TaskSpec:
         return ObjectID(
             self.task_spec.get().ActorCreationDummyObjectId().Binary())
 
+    def previous_actor_task_dummy_object_id(self):
+        """Return the object ID of the previously executed actor task."""
+        if not self.is_actor_task():
+            return ObjectID.nil()
+        return ObjectID(
```
Comment (stephanie-wang) [inline python/ray/includes/task.pxi]: Nit, prefer returning in an `else` block here.

**Review by stephanie-wang — CHANGES_REQUESTED** [review]
A few tiny things, but looks great!

File: src/ray/core_worker/task_interface.cc (hunk 2 above)
Comment (edoakes) [reply src/ray/core_worker/task_interface.cc]: I like this better in general, but it makes the line too long so the linter complains.

Comment (edoakes) [reply src/ray/core_worker/task_interface.cc]: Wait, the line is already too long so the linter complains....

File: python/ray/includes/task.pxi (hunk 3 above)
Comment (edoakes) [reply python/ray/includes/task.pxi]: I'm strongly against returning in an `else` block in these cases, but looks like the usage is mixed in the Python code from a quick glance. We should pick one and stick to it eventually. For now, will leave this as-is to keep consistent with the other near-identical functions.

Comment (stephanie-wang) [reply python/ray/includes/task.pxi]: Returning in `else` is more readable to me, but I don't feel strongly about it.

**Review by stephanie-wang — APPROVED** [review]
Thanks, looks good!

Please try to avoid force-pushing in the future, as it makes it hard to see what the new changes are.

File: src/ray/protobuf/common.proto [hunk 4]
```diff
@@ -86,30 +86,28 @@ message ActorCreationTaskSpec {
...
   // Number of tasks that have been submitted to this actor so far.
   uint64 actor_counter = 5;
-  // If this is an actor task, then this will be populated with all of the new
-  // actor handles that were forked from this handle since the last task on
-  // this handle was submitted.
-  // Note that this is a long string that concatenate all of the new_actor_handle IDs.
+  // This will be populated with all of the new actor handles that were forked
+  // from this handle since the last task on this handle was submitted.
+  // Note that this is a long string that concatenates all of the
+  // new_actor_handle IDs.
```
Comment (raulchen) [inline src/ray/protobuf/common.proto]: `Note that this is a long string that concatenates all of the new_actor_handle IDs.`
This comment no longer is out-dated as we've migrate from fbs to protobuf. Can you help remove it. Thanks.

**Review by raulchen — APPROVED** [review]

**Comment by pcmoritz** [comment]
This is almost ready, there is another Travis error:
```
Use --sandbox_debug to see verbose messages from the sandbox
src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc: In function 'void Java_org_ray_runtime_raylet_RayletClientImpl_nativeSubmitTask(JNIEnv*, jclass, jlong, jbyteArray, jbyteArray)':
src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc:78:76: error: no matching function for call to 'RayletClient::SubmitTask(std::vector<ray::ObjectID>&, ray::TaskSpecification&)'
   auto status = raylet_client->SubmitTask(execution_dependencies, task_spec);
                                                                            ^
In file included from src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc:6:0:
src/ray/raylet/raylet_client.h:85:15: note: candidate: ray::Status RayletClient::SubmitTask(const ray::TaskSpecification&)
   ray::Status SubmitTask(const ray::TaskSpecification &task_spec);
               ^
src/ray/raylet/raylet_client.h:85:15: note:   candidate expects 1 argument, 2 provided
INFO: Elapsed time: 327.299s, Critical Path: 26.18s
INFO: 2021 processes: 1947 remote cache hit, 73 linux-sandbox, 1 worker.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

File: src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc [hunk 5]
```diff
@@ -59,23 +59,17 @@ JNIEXPORT jlong JNICALL Java_org_ray_runtime_raylet_RayletClientImpl_nativeInit(
  * Signature: (J[BLjava/nio/ByteBuffer;II)V
  */
 JNIEXPORT void JNICALL Java_org_ray_runtime_raylet_RayletClientImpl_nativeSubmitTask(
-    JNIEnv *env, jclass, jlong client, jbyteArray cursorId, jbyteArray taskSpec) {
+    JNIEnv *env, jclass, jlong client, jbyteArray taskSpec) {
```
Comment (raulchen) [inline src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc]: It seems that you didn't change the header file. There's an instruction on how to do that in `RayletClientImpl.java`.

**Pull request merged**